### PR TITLE
Default simulation configuration parser

### PR DESCRIPTION
**Pull request by romancardenas**
This first version of the `build_simulation_scenario` function is able to check that the common and default simulation configuration parameters are correct.

The next step is detecting all the entities (i.e., cells) and add them to the output file. To do so, I'll contact @staubibr so I get an idea of how to detect these cells from existing structures.

### CODE REVIEW

File: test/test_simulation.c [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    read_json_file("./data/simulation/2_1_empty_config.json", &json);
+    TEST_ASSERT_EQUAL(SIM_MODEL_SELECTION_INVALID, build_simulation_scenario(json, NULL, NULL));
+
+    read_json_file("./data/simulation/2_2_nonstring_model.json", &json);
+    TEST_ASSERT_EQUAL(SIM_MODEL_SELECTION_INVALID, build_simulation_scenario(json, NULL, NULL));
+
+    read_json_file("./data/simulation/2_3_nonexisting_model.json", &json);
+    TEST_ASSERT_EQUAL(SIM_MODEL_SELECTION_INVALID, build_simulation_scenario(json, NULL, NULL));
+}
+
```
Comment (francoiseblanc) [inline test/test_simulation.c:33]: Only your first test case is executed. I believe that you have to change all your paths to include: ../test/data/simulation

Comment (romancardenas) [reply test/test_simulation.c:33]: Which is your working directory? I selected the `test` folder as the working directory when running the tests. Nonetheless, I'll check the paths as you suggested.

File: test/test_simulation.c [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+void test_invalid_output_config_path(void) {
+    cJSON *json = NULL;
+
+    read_json_file("./data/simulation/5_1_missing_output_config_path.json", &json);
+    TEST_ASSERT_EQUAL(SIM_CONFIG_OUTPUT_PATH_INVALID, build_simulation_scenario(json, NULL, NULL));
+
+    read_json_file("./data/simulation/5_2_invalid_output_config_path.json", &json);
+    TEST_ASSERT_EQUAL(SIM_CONFIG_OUTPUT_PATH_INVALID, build_simulation_scenario(json, NULL, NULL));
+
+    read_json_file("./data/simulation/5_3_existing_output_config_path.json", &json);
```
Comment (francoiseblanc) [inline test/test_simulation.c:75]: I think the path in this .json file must also be changed. Or the error code must be changed.

**Review by francoiseblanc — APPROVED** [review]
Great work! Just a few comments regarding test_simulation.c

**Comment by romancardenas** [comment]
> Great work! Just a few comments regarding test_simulation.c

Thanks, @francoiseblanc ! I've just pushed a new commit addressing your comments. Please, confirm that now the tests run successfully

File: src/simulation_lifecycle/models.c [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+#define MODEL_SIRX_POPULATION "population"
+#define MODEL_SIRX_SUSCEPTIBLE "susceptible"
+#define MODEL_SIRX_INFECTED "infected"
+#define MODEL_SIRX_RECOVERED "recovered"
+#define MODEL_SIRX_DECEASED "deceased"
+
+#define MODEL_SIRX_VIRULENCE "virulence"
+#define MODEL_SIRX_RECOVERY "recovery"
+#define MODEL_SIRX_FATALITY "fatality"
+#define MODEL_SIRX_IMMUNITY "immunity"
```
[the remark this answers is not in the document]
Comment (romancardenas) [reply src/simulation_lifecycle/models.c:31]: I use these macros to check that the specified model in the workflow exists. Also, I check that all the required configuration parameters exist and are of the expected data type (in this case, all are floats). We can remove this and make it less strict. But then simulation may fail.

File: src/simulation_lifecycle/models.c [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    cJSON *c = cJSON_GetObjectItemCaseSensitive(from, MODEL_CELL_CONFIG);
+    if (c == NULL || !cJSON_IsObject(c) || copy_json_values(cJSON_IsNumber, c, default_config,
+                                                            MODEL_SIRX_VIRULENCE, MODEL_SIRX_RECOVERY,
+                                                            MODEL_SIRX_FATALITY, MODEL_SIRX_IMMUNITY, NULL)) {
+        return SIM_MODEL_SPECIFIC_CONFIG_INVALID;
+    }
+    cJSON_AddItemToObject(target, MODEL_CELL_CONFIG, default_state);
+
+    return SUCCESS;
+}
```
[the remark this answers is not in the document]
Comment (romancardenas) [reply src/simulation_lifecycle/models.c:157]: If we do a more generic approach, then the function won't be able to check that the simulation configuration is correct. Thus, the simulation may fail. I'm OK with it, we can discuss this in the next meeting and decide whether to change it.

File: src/simulation_lifecycle/utils/file.c [hunk 5]
```diff
@@ -18,4 +21,38 @@ int write_data_to_file(char *filepath, char *data) {
...
+    fseek(f, 0, SEEK_SET);  /* same as rewind(f); */
+
+    char *string = malloc(fsize + 1);
+    fread(string, 1, fsize, f);
+    fclose(f);
+    string[fsize] = '\0';
+
+    *p_target = cJSON_Parse(string);
+    free(string);
+}
```
[the remark this answers is not in the document]
Comment (romancardenas) [reply src/simulation_lifecycle/utils/file.c:39]: We can merge my branch and when we merge yours, I'll take care of all the collisions (I'm used to working with this Git stuff)

File: src/simulation_lifecycle/models.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (staubibr) [reply src/simulation_lifecycle/models.c:31]: Well you could have the info you need in the simulation section. In addition to the mapping, we would have to add the field type.

I dont mind though. If it adds too much complexity, its not worth it.

File: src/simulation_lifecycle/utils/file.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (staubibr) [reply src/simulation_lifecycle/utils/file.c:39]: I don't mind merging at some point either, I've done it in the past. But I certainly won't say no if you want to do it! I have  an ever growing list of things to do this week...

**Comment by romancardenas** [comment]
@staubibr @francoiseblanc 

I significantly changed the code to address Bruno's comments. Now it is more general. It does not need a specific implementation for each model.
- It checks if there is a binary with the name of the selected model
- It checks that important fields (like `state`) exist, but it cannot tell if it contains valid values. Therefore, that issue will arise when running the simulation.

I think it is a more flexible approach, and even though. What do you think?

**Review by staubibr — APPROVED** [review]
Looks good to me, well done!

**Comment by francoiseblanc** [comment]
Agreed! Really well done!

Is it expected that two of your tests are failing?

 
image

**Review by francoiseblanc — APPROVED** [review]

**Pull request merged**